pkg/controller/kubedirectorcluster/members.go

```diff
@@ -118,6 +118,9 @@ func syncMembers(
 		if _, ok := r.membersByState[memberCreating]; ok {
 			handleCreatingMembers(reqLogger, cr, r, roles, configmetaGenerator)
 		}
+		if _, ok := r.membersByState[memberUpgrading]; ok {
+			handleUpgradingMembers(reqLogger, cr, r, roles)
+		}
```

Comment:
A general thought here...

Modifying the member state machine is potentially pretty tricky to get correct in all cases. It may be something we do need to do in this case, but maybe there's a different way.

When a pod gets upgraded, its container restarts, so KD kicks it back into the create pending state, and then it eventually gets to creating state.

When KD handles a creating-state member, if it has an app setup package then it calls appConfig.

appConfig checks to see if the initial config needs to be run (or re-run, in the case of a previous error). If it does need to run initial config then it runs the startscript with the "--configure" arg (the appPrepConfigRunCmd).

(I think in this case, if for whatever reason we're running initial config, we should not send an upgrade notify to this pod.)

Let's say though that appConfig determines that it does NOT need to run the initial config, because it sees that has alreayd been run on this pod. In that case it returns without doing anything. What it seems like we could do, is modify this part of the process. For example:

If appConfig sees that config has already been run previously, it can check whether the container image has changed (e.g. by referencing that new status data you're adding). If so, then instead of doing nothing, it can run the startscript with the "--upgraded" argument.

Something to think about. If it's possible to avoid adding new member states (with roughly the same or nicer code complexity) then IMO that is preferable.

Reply:
Well, I removed the `memberUpgrading` state and added sending upgrade notification to `appConfig`, as you recommended. Hope, I chose the correct place for it.